test/vm_tests/builder/declaration_parsing_test.dart

```diff
@@ -91,430 +112,1414 @@ main() {
+                expect(versionForSecondMixin.version, Version.v4_mixinBased);
+              });
             });
 
-            test('a props mixin', () {
-              setUpAndParse('''
-                @PropsMixin(keyNamespace: "bar")
-                class FooPropsMixin {}
+            group('and there is a mix of abstract and concrete', () {
+              Iterable<LegacyClassComponentDeclaration> componentDeclarations;
+
+              setUp(() {
+                setUpAndParse('''
+                ${OverReactSrc.abstractState(needsComponent: false, baseName: 'StateFoo').source}
+                
+                ${OverReactSrc.abstractProps(needsComponent: false, baseName: 'PropsFoo').source}
+  
+                ${getBoilerplateString(version: BoilerplateVersions.v2, componentBaseName: 'SecondFoo')}
+                
+                ${getBoilerplateString(version: BoilerplateVersions.v3, componentBaseName: 'ThirdFoo')}
               ''');
-              expect(declarations.propsMixins.single.meta.keyNamespace, 'bar');
+
+                componentDeclarations = declarations.whereType<LegacyClassComponentDeclaration>();
+              });
+
+              test('detects the correct number of components', () {
+                expect(componentDeclarations.length, 2,
+                    reason: 'There should be two different components.');
+              });
+
+              test('with names that match expectations', () {
+                final propsComponentDeclaration =
+                componentDeclarations.firstWhereNameEquals('PropsFoo');
+                final stateComponentDeclaration =
+                componentDeclarations.firstWhereNameEquals('StateFoo');
+                final secondComponent = componentDeclarations.firstWhereNameEquals('SecondFoo');
+                final thirdComponent = componentDeclarations.firstWhereNameEquals('ThirdFoo');
+
+                expect(propsComponentDeclaration, isNull);
+                expect(stateComponentDeclaration, isNull);
+                expect(secondComponent, isNotNull);
+                expect(thirdComponent, isNotNull);
+
+                final abstractProps = declarations.whereType<LegacyAbstractPropsDeclaration>();
+                expect(abstractProps.length, 1);
+                expect(abstractProps.first.props.name.name, r'_$PropsFooProps');
+
+                final abstractState = declarations.whereType<LegacyAbstractStateDeclaration>();
+                expect(abstractState.length, 1);
+                expect(abstractState.first.state.name.name, r'_$StateFooState');
+              });
+
+              test('that are the correct version', () {
+                final versionForFirstMixin =
+                resolveVersion(componentDeclarations.firstWhereNameEquals('SecondFoo').members);
+                final versionForSecondMixin =
+                resolveVersion(componentDeclarations.firstWhereNameEquals('ThirdFoo').members);
```

Comment:
#superFormattingNit It's hard to tell here, but in Webstorm these resolveVersion calls need the two extra tabs